raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java

```diff
@@ -0,0 +1,215 @@
+             ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream)) {
+            objectOutputStream.writeObject(event);
+        } catch (Exception e) {
+            logger.error("Error writing Event to offline storage: " + event.getId(), e);
+        }
+
+        logger.debug(Integer.toString(getNumStoredEvents())
+            + " stored events are now in dir: "
+            + bufferDir.getAbsolutePath());
+    }
+
+    /**
+     * Deletes a buffered {@link Event{ from disk.
+     *
+     * @param event Event to delete from the disk.
+     */
+    @Override
+    public void discard(Event event) {
+        File eventFile = new File(bufferDir, event.getId().toString() + FILE_SUFFIX);
+        if (eventFile.exists()) {
+            logger.debug("Discarding Event from offline storage: " + eventFile.getAbsolutePath());
+            eventFile.delete();
+        }
+    }
+
+    /**
+     * Attempts to open and deserialize a single {@link Event} from a {@link File}.
+     *
+     * @param eventFile File to deserialize into an Event
+     * @return Event from the File, or null
+     */
+    private Event fileToEvent(File eventFile) {
+        FileInputStream fileInputStream;
+        Object eventObj;
+
+        try {
+            fileInputStream = new FileInputStream(new File(eventFile.getAbsolutePath()));
+            ObjectInputStream ois = new ObjectInputStream(fileInputStream);
+            eventObj = ois.readObject();
+        } catch (Exception e) {
+            logger.error("Error reading Event file: " + eventFile.getAbsolutePath(), e);
+            eventFile.delete();
+            return null;
+        }
+
+        try {
+            return (Event) eventObj;
+        } catch (Exception e) {
+            logger.error("Error casting Object to Event: " + eventFile.getAbsolutePath(), e);
+            eventFile.delete();
+            return null;
+        }
+    }
+
+    /**
+     * Returns the next *valid* {@link Event} found in an Iterator of Files.
+     *
+     * @param files Iterator of Files to deserialize
+     * @return The next Event found, or null if there are none
+     */
+    private Event getNextEvent(Iterator<File> files) {
+        while (files.hasNext()) {
+            File file = files.next();
+
+            // only consider files that end with FILE_SUFFIX
+            if (!file.getAbsolutePath().endsWith(FILE_SUFFIX)) {
+                continue;
+            }
+
+            Event event = fileToEvent(file);
+            if (event != null) {
+                return event;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns an Iterator of Events that are stored to disk at the point in time this method
+     * is called. Note that files may not deserialize correctly, may be corrupted,
+     * or may be missing by the time we attempt to open them - so some care is taken to
+     * only return valid {@link Event}s.
```

Comment:
That makes sense on second reading. 😅  Might help to make this a little more explicit for bad readers like myself.